### ISSUE TITLE
write_all hangs after read on a pty slave

### ISSUE THREAD

**Issue by russell-islam** [tokio-rs/tokio#5053]
**Version**
tokio v1.19.2 (*)

**Platform**
x86_64 x86_64 GNU/Linux

**Description**

I tried to read the data from a PTY slave and write the user input back to the pty slave. But it hangs during the first input write to the file

I tried this code: Entry point run_stream, hanging function call  file.write_all(&[*b]).await?;

enum EventLoopResult {
    // Input received
    Input(usize),
    // Output received
    Output(usize),
    // An error occurred
    Error(anyhow::Error),
}

// Run an interactive console session of pty slave
pub(crate) async fn run_stream(file_name: &str) -> Result<()> {

    let mut file = OpenOptions::new()
        .write(true)
        .read(true)
        .create(true)
        .open(file_name)
        .await?;

    let mut stdin_stream = stdin();
    let mut stdout_stream = stdout();

    let mut console_buf = [0u8; 256];
    let mut stdin_buf = [0u8; 256];

    loop {
        let loop_result = select! {
            bytes_read = file.read(&mut console_buf[..]) => {
                match bytes_read {
                    Err(e) => EventLoopResult::Error(e.into()),
                    Ok(b) => EventLoopResult::Output(b),
                }
            },
            bytes_read = stdin_stream.read(&mut stdin_buf[..]) => {
                match bytes_read {
                    Err(e) => EventLoopResult::Error(e.into()),
                    Ok(b) => EventLoopResult::Input(b),
                }
            }
        };

        // Process results
        match loop_result {
            EventLoopResult::Error(e) => {
                return Err(e);
            }
            EventLoopResult::Input(bytes_read) => {
                println!("Bytes read: {}", bytes_read);
                if process_input(&stdin_buf[0..bytes_read], &mut file).await? {
                    // Requested shutdown
                    break;
                }
            }
            EventLoopResult::Output(bytes_read) => {
                stdout_stream.write_all(&console_buf[0..bytes_read]).await?;
                stdout_stream.flush().await?;
            }
        };
    }

    Ok(())
}

/// Process input data
async fn process_data<T>(buffer: &[u8],  file: &mut T) -> Result<bool>
where
    T: AsyncWriteExt + Unpin,
{

    file.write_all(&[*b]).await?;

    Ok(true)
}

I expected to see this happen: Should continue the read and write

Instead, this happened: Blocked on write_all

**Comment by Darksonn** [comment]
Please modify your issue to properly use code blocks.

**Comment by russell-islam** [comment]
> Please modify your issue to properly use code blocks.

How? Sorry!!

**Comment by Darksonn** [comment]
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Comment by russell-islam** [comment]
> https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

Done. Thank you. Do you have any suggestion to the above code?

**Comment by Noah-Kennedy** [comment]
You can also do \`\`\`rust to make syntax highlighting work.

**Comment by Noah-Kennedy** [comment]
Thanks!

**Comment by Noah-Kennedy** [comment]
Is anything consuming the data written in `write_all`? It's possible that `write_all` blocks permanently because nothing is reading from the other end.

**Comment by russell-islam** [comment]
> Is anything consuming the data written in `write_all`? It's possible that `write_all` blocks permanently because nothing is reading from the other end.

Same PTY pair works fine with socat.

**Comment by russell-islam** [comment]
I also tried setting O_NDELAY/O_NONBLOCK by custom flags during opening the file. If I set either one of these two files read gives me EAGAIN infinitely.

**Comment by Darksonn** [comment]
I don't know what's wrong with your code, but I generally believe that any kind of interactive terminal IO should happen outside of Tokio on dedicated threads. Would that work for your code?

**Comment by russell-islam** [comment]
Yes. I was able to do it using epoll. But tokio should work. right?
Same code base works fine for tcpstream(tokio). File has issue.

**Comment by russell-islam** [comment]
If this is something tokio does not support, I can close the bug.

**Comment by Noah-Kennedy** [comment]
@russell-islam tokio::fs is meant for filesystem operations, which obviously don't work with epoll, so it outsources work to a thread pool. I'd try using std::fs::OpenOptions and wrapping the file in AsyncFd, to watch it with tokio's epoll instance.

**Comment by russell-islam** [comment]
For my use case, I did not use tokio at all. I have used std FS OpenOptions with O_NBLOCK then used epoll in  a loop. And it worked with PTY slave. 

The above code base works fine if I use TCP Stream instead of PTY salve as the source of the data read/write. That means tokio works fine with TcpStream but not with file. I wanted to know if this is expected behavior. If so I can close the bug.

**Comment by Noah-Kennedy** [comment]
@russell-islam What I am suggesting is to do what you were doing with `std::fs::OpenOptions` and O_NONBLOCK, but to use AsyncFd instead of epoll. AsyncFd basically just lets you `.await` epoll readiness by hooking into tokio's epoll driver.

**Comment by sharifhsn** [comment]
Can be closed, #6467 clarifies in the [`tokio::fs` docs](https://docs.rs/tokio/latest/tokio/fs/index.html) that it should not be used for special files like a PTY slave.

> The `tokio::fs` module should only be used for ordinary files. Trying to use it with e.g., a named pipe on Linux can result in surprising behavior, such as hangs during runtime shutdown. For special files, you should use a dedicated type such as [`tokio::net::unix::pipe`](https://docs.rs/tokio/latest/tokio/net/unix/pipe/index.html "mod tokio::net::unix::pipe") or [`AsyncFd`](https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html "struct tokio::io::unix::AsyncFd") instead.

For the specific use case that @russell-islam posted here, there are two third-party libraries that provide async `tokio` support for PTY slaves.

`tokio-fd`: the most direct solution in the original example, a version of `AsyncFd` that is specialized for non-blocking use of Unix file descriptors

```diff
-    let mut file = OpenOptions::new()
-        .write(true)
-        .read(true)
-        .create(true)
-        .open(file_name)
-        .await?;
+    let fd = std::fs::OpenOptions::new().read(true).write(true).open(file_name)?.into_raw_fd();
+    let mut file = tokio_fd::AsyncFd::try_from(fd)?;   // AsyncRead + AsyncWrite; sets O_NONBLOCK
```

`pty-process`: more maintained and provides a richer interface. Requires `unsafe` to construct from arbitrary fd, provides safe methods to construct your own PTY

```diff
-    let mut file = OpenOptions::new()
-        .write(true).read(true).create(true)
-        .open(file_name).await?;
+    let f = std::fs::OpenOptions::new()
+        .read(true).write(true)
+        .custom_flags(libc::O_NONBLOCK)
+        .open(file_name)?;
+    let mut file = unsafe { pty_process::Pty::from_fd(OwnedFd::from(f))? }; // I/O safe, assumes file belongs to PTY and is non-blocking
```
